### PR TITLE
Implement Rare Feature Encoder

### PR DESCRIPTION
**Pull request by ibenfjord**
Implemented _RareFeatureEncoder_ Transformer which is able to replaces categories with a chosen value, if they occur less than a chosen threshold.

Obs!:
- Nan values are not included, they should be transformed before.
- The threshold can be an int or percentage.
- Can be used in cross validation and gridsearch.
- Can handle columns which are either not in Train or Test.
- Column name needs to be identical.

- [x] close issue 498
- [x] Tests added
- [x] Passes flake8
- [x] Updated CHANGELOG

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,7 @@ human friendly manner
 - Added a missing data visualization to `Dataset.plot`
 - FillNA now accepts a `is_nan`flag which adds a flag indicating that a value was missing
 - `Model.make_prediction` now accepts a `use_cache`flag to score everything in cached `.x`
+- Added Tranformer `RareFeatureEncoder`
```
Comment (andersbogsnes) [inline CHANGELOG.md:11]:
```suggestion
- Added a new Transformer:  `RareFeatureEncoder`
```

File: src/ml_tooling/transformers/rare_feature.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+class RareFeatureEncoder(BaseEstimator, TransformerMixin):
+    """
+    Replace values which occurs below the chosen threshold.
+    The threshold can be a procent or int value.
+    The values are replaced with the chosen fill value.
+    Column names need to be the same.
+    """
+
+    def __init__(self, threshold=0.2, fill_rare="Rare"):
```
Comment (andersbogsnes) [inline src/ml_tooling/transformers/rare_feature.py:13]: Remember your type annotations:

```suggestion
    def __init__(self, threshold: Union[int, float] = 0.2, fill_rare: Any = "Rare"):
```

Comment (andersbogsnes) [reply src/ml_tooling/transformers/rare_feature.py:13]: Missing docstring for the `__init__`

File: src/ml_tooling/transformers/rare_feature.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
+import pandas as pd
+from sklearn.base import BaseEstimator, TransformerMixin
+
+
+class RareFeatureEncoder(BaseEstimator, TransformerMixin):
+    """
+    Replace values which occurs below the chosen threshold.
```
Comment (andersbogsnes) [inline src/ml_tooling/transformers/rare_feature.py:7]: Try to write the docstring more like a description instead of bullets - see some of the other transformers for inspiration

File: src/ml_tooling/transformers/rare_feature.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    Column names need to be the same.
+    """
+
+    def __init__(self, threshold=0.2, fill_rare="Rare"):
+        self.threshold = threshold
+        self.fill_rare = fill_rare
+        self.mask_dict = {}
+
+    def fit(self, X: pd.DataFrame, y=None):
+        for col in X.columns:
```
Comment (andersbogsnes) [inline src/ml_tooling/transformers/rare_feature.py:19]: A minor improvement:
```
normalize = True if self.threshold < 1 else False
frequencies = X[col].value_counts(normalize=normalize)
```

File: tests/test_transformers.py [hunk 5]
```diff
@@ -763,3 +764,118 @@ def test_binarize_works_in_gridsearch(self, train_iris_dataset):
         model = Model(grid)
         result = model.score_estimator(train_iris_dataset)
         assert isinstance(result, Result)
+
+
+class TestRareFeatureEncoder(TransformerBase):
+    def test_rare_feature_encoder_returns_correctly_series(self):
+        threshold = 2
```
Comment (andersbogsnes) [inline tests/test_transformers.py:771]: Don't need to declare them here - just use them directly in the RareFeatureEncoder instance

File: tests/test_transformers.py [hunk 6]
```diff
@@ -763,3 +764,118 @@ def test_binarize_works_in_gridsearch(self, train_iris_dataset):
...
+        result = rare.transform(new_data)
+
+        test_data = pd.DataFrame(
+            {
+                "categorical": [99, 99, 99, "a", 99, 99, 3],
+                "numerical": [1, 2, 3, 4, 5, 6, 7],
+            }
+        )
+
+        assert isinstance(result, pd.DataFrame)
```
Comment (andersbogsnes) [inline tests/test_transformers.py:810]: You can use `pd.testing.assert_frame_is_equal` here to avoid multiple asserts

File: tests/test_transformers.py [hunk 7]
```diff
@@ -763,3 +764,118 @@ def test_binarize_works_in_gridsearch(self, train_iris_dataset):
...
+        assert len(new_data) == len(result)
+        assert test_data.equals(result)
+
+    def test_rare_feature_encoder_handle_nan(self):
```
Comment (andersbogsnes) [inline tests/test_transformers.py:814]: For test code it is fine to have long, descriptive function names - we only see them in the test output and when reading them, so the more descriptive, the better. So maybe something along the lines of:
```suggestion
    def test_rare_feature_encoder_doesnt_count_nans(self):
```
```

File: tests/test_transformers.py [hunk 8]
```diff
@@ -763,3 +764,118 @@ def test_binarize_works_in_gridsearch(self, train_iris_dataset):
...
+                "categorical": ["Rare", "Rare", "Rare", "a", "Rare", "Rare", np.nan],
+                "numerical": [1, 2, 3, 4, 5, 6, 7],
+            }
+        )
+
+        assert isinstance(result, pd.DataFrame)
+        assert new_data.shape == result.shape
+        assert test_data.equals(result)
+
+    def test_rare_feature_encoder_use_procent(self):
```
Comment (andersbogsnes) [inline tests/test_transformers.py:841]:
```suggestion
    def test_rare_feature_encoder_correctly_counts_rare_when_given_percent_threshold(self):
```

File: src/ml_tooling/transformers/rare_feature.py (hunk 4 above)
Comment (andersbogsnes) [reply src/ml_tooling/transformers/rare_feature.py:19]: I'm pretty sure we have an `is_percent` utility function to test if something is a percentage - better to use that

File: src/ml_tooling/transformers/rare_feature.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+            if self.threshold < 1:
+                frequencies = X[col].value_counts(normalize=True)
+            else:
+                frequencies = X[col].value_counts()
+
+            condition = frequencies < self.threshold
+            mask_obs = frequencies[condition].index
+            self.mask_dict[col] = dict.fromkeys(mask_obs, self.fill_rare)
```
Comment (andersbogsnes) [inline src/ml_tooling/transformers/rare_feature.py:27]: Nice use of `dict.fromkeys` :+1:

File: src/ml_tooling/transformers/rare_feature.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+    def fit(self, X: pd.DataFrame, y=None):
+        for col in X.columns:
+            if self.threshold < 1:
+                frequencies = X[col].value_counts(normalize=True)
+            else:
+                frequencies = X[col].value_counts()
+
+            condition = frequencies < self.threshold
+            mask_obs = frequencies[condition].index
```
Comment (andersbogsnes) [inline src/ml_tooling/transformers/rare_feature.py:26]: Would probably combine these two, since we don't refer to it afterwards

```suggestion
            mask_obs = frequencies[frequencies < self.threhold].index
```

File: tests/test_transformers.py [hunk 11]
```diff
@@ -763,3 +764,118 @@ def test_binarize_works_in_gridsearch(self, train_iris_dataset):
         model = Model(grid)
         result = model.score_estimator(train_iris_dataset)
         assert isinstance(result, Result)
+
+
+class TestRareFeatureEncoder(TransformerBase):
+    def test_rare_feature_encoder_returns_correctly_series(self):
```
Comment (andersbogsnes) [inline tests/test_transformers.py:770]: There is no Series being used here, so the function name is a bit misleading

File: tests/test_transformers.py [hunk 12]
```diff
@@ -763,3 +764,118 @@ def test_binarize_works_in_gridsearch(self, train_iris_dataset):
...
+        fill_rare = "Rare"
+        rare = RareFeatureEncoder(threshold=threshold, fill_rare=fill_rare)
+
+        data = pd.DataFrame({"categorical": [1, "a", "a", 2, "b", 1]})
+        rare.fit(data)
+
+        new_data = pd.DataFrame({"categorical": [1, 1, 1, "a", "b", "b", 3]})
+        result = rare.transform(new_data)
+
+        test_data = pd.DataFrame({"categorical": [1, 1, 1, "a", "Rare", "Rare", 3]})
```
Comment (andersbogsnes) [inline tests/test_transformers.py:781]: would probably rename this to `expected` - in general, we want the pattern to be `assert result == expected`

File: tests/test_transformers.py [hunk 13]
```diff
@@ -763,3 +764,118 @@ def test_binarize_works_in_gridsearch(self, train_iris_dataset):
...
+        )
+
+        assert isinstance(result, pd.DataFrame)
+        assert test_data.shape == result.shape
+        assert test_data.equals(result)
+
+    def test_rare_feature_encoder_can_be_used_cv(self, train_iris_dataset):
+        threshold = 0.2
+        fill_rare = "Rare"
+        rare = RareFeatureEncoder(threshold=threshold, fill_rare=fill_rare)
```
Comment (andersbogsnes) [inline tests/test_transformers.py:868]: Creating a RareFeatureEncoder could be pulled into a fixture in the class, since it's used a few times

File: tests/test_transformers.py [hunk 14]
```diff
@@ -763,3 +764,118 @@ def test_binarize_works_in_gridsearch(self, train_iris_dataset):
...
+                "categorical": ["Rare", "a", "Rare", "Rare", np.nan],
+                "numerical": [1, 2, 3, 4, 5],
+            }
+        )
+
+        assert isinstance(result, pd.DataFrame)
+        assert test_data.shape == result.shape
+        assert test_data.equals(result)
+
+    def test_rare_feature_encoder_can_be_used_cv(self, train_iris_dataset):
```
Comment (andersbogsnes) [inline tests/test_transformers.py:865]: Annotations:
```suggestion
    def test_rare_feature_encoder_can_be_used_cv(self, train_iris_dataset: pd.DataFrame):
```

File: src/ml_tooling/transformers/rare_feature.py (hunk 3 above)
Comment (ibenfjord) [reply src/ml_tooling/transformers/rare_feature.py:7]: Instead?: 
Replaces categories with a value, if they occur less than a threshold.

File: src/ml_tooling/transformers/rare_feature.py (hunk 4 above)
Comment (ibenfjord) [reply src/ml_tooling/transformers/rare_feature.py:19]: It have been changed, so it uses "is_percent"

File: tests/fixtures/estimators.py [hunk 15]
```diff
@@ -123,3 +128,8 @@ def tmp_open_estimator_pickle(path: Optional[pathlib.Path] = None):
         return f
 
     return tmp_open_estimator_pickle
+
+
+@pytest.fixture()
```
Comment (andersbogsnes) [inline tests/fixtures/estimators.py:133]: Since this fixture is only used in the TestRareFeatureEncoder class, it would be better to have the fixture defined in that class - then it can also be class scoped

File: tests/fixtures/dataframes.py [hunk 16]
```diff
@@ -10,6 +10,11 @@ def categorical():
     )
 
 
+@pytest.fixture
```
Comment (andersbogsnes) [inline tests/fixtures/dataframes.py:13]: If this fixture is only used in the Test class, then I would prefer to move it there

File: src/ml_tooling/transformers/rare_feature.py [hunk 17]
```diff
@@ -1,29 +1,25 @@
...
 from sklearn.base import BaseEstimator, TransformerMixin
 
 
 class RareFeatureEncoder(BaseEstimator, TransformerMixin):
     """
-    Replace values which occurs below the chosen threshold.
-    The threshold can be a procent or int value.
-    The values are replaced with the chosen fill value.
-    Column names need to be the same.
+    Replaces categories with a value, if they occur less than a threshold.
```
Comment (andersbogsnes) [inline src/ml_tooling/transformers/rare_feature.py:9]:
```suggestion
    Replaces categories with a specified value, if they occur less often than the provided threshold.
```

File: src/ml_tooling/transformers/rare_feature.py [hunk 18]
```diff
@@ -1,29 +1,25 @@
...
     """
 
-    def __init__(self, threshold=0.2, fill_rare="Rare"):
+    def __init__(self, threshold: Union[int, float] = 0.2, fill_rare: Any = "Rare"):
```
Comment (andersbogsnes) [inline src/ml_tooling/transformers/rare_feature.py:12]: Missing an `__init__` docstring explaining the parameters

File: src/ml_tooling/transformers/rare_feature.py [hunk 19]
```diff
@@ -1,29 +1,25 @@
...
         self.mask_dict = {}
 
     def fit(self, X: pd.DataFrame, y=None):
         for col in X.columns:
-            if self.threshold < 1:
-                frequencies = X[col].value_counts(normalize=True)
-            else:
-                frequencies = X[col].value_counts()
+            normalize = _is_percent(self.threshold)
+            frequencies = X[col].value_counts(normalize=normalize)
```
Comment (andersbogsnes) [inline src/ml_tooling/transformers/rare_feature.py:20]: Can oneliner this:
```suggestion
            frequencies = X[col].value_counts(normalize=_is_percent(self.threshold))
```

File: src/ml_tooling/transformers/rare_feature.py [hunk 20]
```diff
@@ -1,29 +1,25 @@
...
 
-            condition = frequencies < self.threshold
-            mask_obs = frequencies[condition].index
+            mask_obs = frequencies[frequencies < self.threshold].index
```
Comment (andersbogsnes) [inline src/ml_tooling/transformers/rare_feature.py:22]: We can make this variable name more informative - maybe something like this?
```suggestion
            rare_values = frequencies[frequencies < self.threshold].index
```

File: src/ml_tooling/transformers/rare_feature.py [hunk 21]
```diff
@@ -1,29 +1,25 @@
...
-    The values are replaced with the chosen fill value.
-    Column names need to be the same.
+    Replaces categories with a value, if they occur less than a threshold.
     """
 
-    def __init__(self, threshold=0.2, fill_rare="Rare"):
+    def __init__(self, threshold: Union[int, float] = 0.2, fill_rare: Any = "Rare"):
         self.threshold = threshold
         self.fill_rare = fill_rare
         self.mask_dict = {}
```
Comment (andersbogsnes) [inline src/ml_tooling/transformers/rare_feature.py:15]: We can make this more informative - something like `rare_values`?

File: tests/test_transformers.py [hunk 22]
```diff
@@ -767,30 +767,10 @@ def test_binarize_works_in_gridsearch(self, train_iris_dataset):
...
-        assert test_data.equals(result)
-
-    def test_rare_feature_encoder_returns_correctly_dataframe(self):
-        threshold = 2
-        fill_rare = 99
-        rare = RareFeatureEncoder(threshold=threshold, fill_rare=fill_rare)
-
-        data = pd.DataFrame({"categorical": [1, "a", "a", 2, "b", 2]})
-        rare.fit(data)
+    def test_rare_feature_encoder_returns_correctly_when_given_one_column_dataframe(
```
Comment (andersbogsnes) [inline tests/test_transformers.py:770]: "Correctly" is a bit vague here.

Now it is testing if the transformed data and the input data have the same shape - maybe we can specify that in the function name?

File: tests/test_transformers.py [hunk 23]
```diff
@@ -799,82 +779,75 @@ def test_rare_feature_encoder_returns_correctly_dataframe(self):
...
-
-        assert isinstance(result, pd.DataFrame)
-        assert len(new_data) == len(result)
-        assert test_data.equals(result)
-
-    def test_rare_feature_encoder_handle_nan(self):
-        threshold = 2
-        fill_rare = "Rare"
-        rare = RareFeatureEncoder(threshold=threshold, fill_rare=fill_rare)
+    def test_rare_feature_encoder_returns_correctly_dataframe(
```
Comment (andersbogsnes) [inline tests/test_transformers.py:784]: Here returns correctly is more apt, since it's checking against a known, correct DataFrame :+1:

File: src/ml_tooling/transformers/rare_feature.py [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+class RareFeatureEncoder(BaseEstimator, TransformerMixin):
+    def __init__(self, threshold: Union[int, float] = 0.2, fill_rare: Any = "Rare"):
+        """
+        Replaces categories with a specified value, if they occur less often than
+        the provided threshold.
+
+        Parameters
+        ----------
+        threshold: int, float
+            Threshold of a categorie minimum occurrence.
```
Comment (andersbogsnes) [inline src/ml_tooling/transformers/rare_feature.py:16]: Maybe a bit more sentencey?

"Sets the threshold for when a value is considered rare. Any value which occurs less than (or equal?) to the threshold will be replaced with `fill_rare`. If threshold is a float, it will be considered a percentage and if it is an int, `threshold` will be considered the minimum number of observations"

File: src/ml_tooling/transformers/rare_feature.py [hunk 25]
```diff
@@ -0,0 +1,41 @@
...
+            Threshold = float: Percent.
+            Threshold = Int: Number.
+        fill_rare: Any
+            Value to use to fill categorie.
```
Comment (andersbogsnes) [inline src/ml_tooling/transformers/rare_feature.py:20]:
```suggestion
            Fill value to use when replacing rare categories
```
